fix spaceless deprecation

garak opened the pull request:
Fix #395

stof commented on .github/workflows/build.yaml:
```diff
@@ -4,6 +4,9 @@ on:
     push: ~
 
+env:
+    SYMFONY_DEPRECATIONS_HELPER: max[self]=0
```
I would not add this globally. The existing behavior is that other jobs report *all* deprecations as being an issue, and that's the goal. We want the CI to report direct deprecations (to avoid our users complaining that KnpMenu triggers deprecations when using it because our CI would not have warned us).

The `-prefer-lowest` job uses a special config restricting only our own deprecations and not direct or indirect ones, because we cannot expect the lowest dependency version to be deprecation-free.

stof commented on .github/workflows/build.yaml:
```diff
@@ -49,15 +52,16 @@ jobs:
+                      php: '8.3'
                       dev: true
-                      symfony: 7.0.*
+                      symfony: 7.1.*
```
the job using dev deps should not force a Symfony version at all (especially given that this is *not* the dev version of Symfony)

stof commented on src/Knp/Menu/Twig/MenuExtension.php:
```diff
@@ -36,6 +36,7 @@ public function getFilters(): array
     {
         return [
             new TwigFilter('knp_menu_as_string', [$this, 'pathAsString']),
+            new TwigFilter('knp_menu_spaceless', [$this, 'spaceless'], ['is_safe' => ['html']]),
```
Shouldn't we rather deprecate our `compressed_root` block instead of providing our own spaceless filter suffering from the same shortcomings than the Twig one that has been deprecated because of them ?

garak replied:
My initial idea was just about making out tests not failing about the twig deprecation, but I didn't manage to get it (using the environment variable SYMFONY_DEPRECATIONS_HELPER didn't work).
If we deprecated our internal block, we'll end up with the same problem (tests failing because of the deprecation)

tacman replied:
There's certainly room to take advantage of newer Symfony technologies in this package, although even moreso in the bundle.  Like replacing the macros with Twig Components or even some twig helper functions/filters .

tacman commented:
Any chance this can be merged and a new release can be published?  Twig 4 will be out soon, better to get ahead of this.  Every one of my sites uses this wonderful library, so I see this message all the time.
```
 c:c --env=prod

 // Clearing the cache for the prod environment with debug false                                                        

[2025-02-06T15:19:57.084445+00:00] deprecation.INFO: User Deprecated: Since twig/twig 3.12: Twig Filter "spaceless" is deprecated in /home/tac/g/sites/member-directory/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views/knp_menu.html.twig at line 12. {"exception":"[object] (ErrorException(code: 0): User Deprecated: Since twig/twig 3.12: Twig Filter \"spaceless\" is deprecated in /home/tac/g/sites/member-directory/vendor/knplabs/knp-menu/src/Knp/Menu/Resources/views/knp_menu.html.twig at line 12. at /home/tac/g/sites/member-directory/vendor/twig/twig/src/DeprecatedCallableInfo.php:65)"} []
                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully cleared.                                          
```

garak commented:
I rebased this PR and updated it, now the CI passes without deprecations.

The question raised by @stof is still open, though. This update would formally allow the library to be used with Twig 4, but the possible problem with the spaceless filter is not solved.

ro0NL commented:
maybe im missing it, but why not just remove the deprecated spaceless function from the template here? basically what #400 does.

curry684 commented:
> maybe im missing it, but why not just remove the deprecated spaceless function from the template here? basically what https://github.com/KnpLabs/KnpMenu/pull/400 does.

The "problem" is that this very library is more or less the prototype case of why the `spaceless` filter was created in the first place - nested unordered lists have of old had the issue that making the HTML readable would often result in stray whitespace causing layout issues. For example:
```html
<ul>
   <li>Aap</li>
   <li>Noot</li>
   <li>Mies</li>
</ul>
```
Coupled with CSS rule `li { display: inline-block; }` will render differently with `spaceless` as it will remove the 5 to 8 pixel gap between elements resulting from the whitespace in between collapsing to a single space in the current font.

In other words: just removing the `spaceless` from the root template will likely break many, *MANY*, layouts using this library.

ro0NL commented:
:sweat_smile: ok, i dont want to find out. Then id say port the twig spaceless filter as is here.

curry684 commented:
Here's a Codepen illustrating the problem: https://codepen.io/niels-keurentjes/pen/ByaKrrQ

ro0NL commented on src/Knp/Menu/Twig/MenuExtension.php:
```diff
@@ -121,4 +122,9 @@ public function isAncestor(ItemInterface $item, ?int $depth = null): bool
+
+    public function spaceless(string $content): string
+    {
+        return trim((string) preg_replace('/>\s+</', '><', $content));
```
https://github.com/twigphp/Twig/blob/4effb660b3eb6a2caa10fce0f88f06bdac0a4fd6/src/Extension/CoreExtension.php#L1233

garak replied:
I don't see the reason for that. `$content` is already a string, so the null coalescing is unnecessary.  We need to cast the result instead (because preg_replace could return a null value), which is not done in Twig source code.

ro0NL replied:
it's fine. It's merely a reference to the original function for future readers.

curry684 commented:
>  Then i'd say port the twig spaceless filter as-is here.

I am tempted to agree as in practice `spaceless` is a useless filter commonly only used for all the wrong reasons (obfuscation and compression) except for the single exceptional case of nested menu structures. For now I think that grants KnpMenu the 'right' to maintain the legacy implementation until such time as a better replacement is made or we deem it unnecessary to provide the BC.

stof commented on src/Knp/Menu/Twig/MenuExtension.php:
```diff
@@ -121,4 +122,9 @@ public function isAncestor(ItemInterface $item, ?int $depth = null): bool
         return $this->runtimeExtension->isAncestor($item, $depth);
     }
+
+    public function spaceless(string $content): string
```
I suggest making this a static method instead, which will allow the compiled template to call it directly.

And I would definitely tag the implementation method as `@internal`

wouterj commented:
The issue with `spaceless` is not that it is used for "wrong reasons". The problem is it has no knowledge of block or inline nodes, and ignorantly removes all whitespace everywhere between tags.

I wouldn't be surprised if that means that having a bold tag in part of a menu item label will be broken currently.

Wouldn't it be a better option to add a `menu_spaceless` extension that exclusively removes whitespace between `<li>` elements?

ro0NL commented:
solving the deprecation in a BC manner is a 1st concern to me
then if something is currently broken, that's a 2nd concern to me

curry684 commented:
Exactly, it doesn't matter if right now it's breaking bold tags and everything, because at least now everybody's counting on it to be broken. We just want it to remain reliably broken, without the deprecation, and with forward compatibility with Twig 4, whilst we ponder whether a new major of this library should fix the underlying issue differently.

The currently proposed fix can be shipped as patch release within Semver constraints. Let's do that and open a new central issue discussing the future.

stof commented:
For the next major version of that library, I think we should be relying on whitespace control to remove Twig indentation spaces from the output (and this will have no runtime cost as the removal is even done at compile time).
This will impact projects doing custom themes, as they will need to make sure to remove spaces in any place where the removal is important.

curry684 commented:
Yes I looked into a more fundamental solution today, but indeed that is the only way to go, and I don't see any way to do it in a backwards compatible way.

ro0NL commented:
maybe there's a css fix for future major upgrades, so we can stop rely on spacing in output

stof commented:
@ro0NL if you rely on `display: inline-block`, the difference between no spaces and some spaces is relevant by design of how inline flow works (otherwise, words would not have spacing between them). If you use other kind of layout for the menu (`display: flex` on the `ul` for instance), they make no difference. But knp-menu does not manage the styling of the menu.

curry684 commented:
> maybe there's a css fix for future major upgrades, so we can stop rely on spacing in output

The CSS fix is essentially a matter of removing the spaceless filter, fixing the built-in templates to use whitespace control tags, slapping a v4 major version number on it, and adding a large bold section to the release notes stating "B/C break: all your template overrides are now broken if you are relying on inline block rendering, which is what older Bootstrap versions have done for years".

We cannot fix for others how they use the library, so we have to treat it as a B/C break for every user. And if we were to do this, I'd recommend updating a ton of other stuff as well.

stof commented:
@curry684 I fail to see how this is a **CSS** fix

curry684 commented:
It isn't  😆    The implication is that, to fix it in CSS, is to force every library user to revise their local CSS. We cannot fix it here as we explained.
